syntax = "proto3";

package genx;

service CoreService {
  // Data Preparation Service
  rpc Ingest(IngestRequest) returns (JobResponse) {}

  // Embedding Service
  rpc GenerateEmbedding(EmbeddingRequest) returns (EmbeddingResponse) {}
  rpc QueryEmbedding(QueryRequest) returns (QueryResponse) {}
}

message IngestRequest {
  string job_id = 1;
  string framework = 2;
  map<string, string> config = 3;
  InputDataConfig input_data_config = 4;
  map<string, string> loader_params = 5;
}

message InputDataConfig {
  string type = 1;
  string fileNames = 2;
  string dbName = 3;
  string collectionName = 4;
}

message JobResponse {
  string job_id = 1;
  string status = 2;
  string activity_type = 3;
  string error = 4;
}

message EmbeddingRequest {
  string job_id = 1;
  string model = 2;
  repeated string texts = 3;
  string collection_id = 4;
  map<string, string> metadata = 5;
}

message EmbeddingResponse {
  string job_id = 1;
  string status = 2;
  repeated float embeddings = 3;
  string collection_id = 4;
  string error = 5;
}

message QueryRequest {
  string query_id = 1;
  string model = 2;
  string query_text = 3;
  string collection_id = 4;
  int32 top_k = 5;
}

message QueryResponse {
  string query_id = 1;
  string status = 2;
  repeated Result results = 3;
  string error = 4;
}

message Result {
  string document_id = 1;
  float score = 2;
  map<string, string> metadata = 3;
}